### PR TITLE
Stacy's Media Ranker submission

### PR DESCRIPTION
**Pull request by StacyLundquist**
# Media Ranker
Congratulations! You're submitting your assignment!

## Reflection//Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the work model I created a (currently non-working) method to change the vote button to a remove vote if the person had already voted.
Describe how you approached testing that model method. What edge cases did you come up with? | The idea was to test what happens when the user isn't logged in, or when the vote is nil.
What are `session` and `flash`? What is the difference between them? | Session is the browsing session, and is connected to a particular user. Flash is a message that lasts for a cycle.
What was one thing that you gained more clarity on through this assignment? |  Ummmm, I gained more confusion of a lot of things! I gained clarity on exactly how much coffee I can drink in a day and how to generate models and controllers.
What is the Heroku URL of your deployed application? | http://adiec14-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

### CODE REVIEW

File: test/models/work_test.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      expect(work).must_respond_to check
+    end
+  end
+
+  describe "relationships" do
+    # Your tests go here
+  end
+
+  describe "validations" do
+    let work = works(:test_book)
```
Comment (CheezItMan) [inline test/models/work_test.rb]: This isn't how a `let` is written

```suggestion
    let (:work) { works(:test_book) }
```

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      # Assert
+      must_respond_with :success
+    end
+  end
+
+  describe "new" do
+    it "can get the new user" do
+
+      # Act
+      get new_user_path
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb]: This test is failing because in `routes.rb` you don't have the route given a nickname.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'works#home'
+
+  resources :works do
+    resources :votes, only: [:create, :destroy]
+  end
+
+  resources :users, only: [:index, :show]
+  get '/signup', to: 'users#new'
```
Comment (CheezItMan) [inline config/routes.rb]:
```suggestion
  get '/signup', to: 'users#new', as: "new_user"
```

File: test/controllers/users_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+      # Assert
+      must_respond_with :success
+    end
+  end
+
+  describe "show" do
+    it "can get a valid user" do
+          # Act
+          get user_path(first.id)
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb]: Missing The model class

```suggestion
      get user_path(User.first.id)
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  def index
+    @users = User.all.order(:created_at)
+  end
+
+  def new
+    @user = User.new()
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
```
Comment (CheezItMan) [inline app/controllers/users_controller.rb:12]: You need to do something if the user doesn't exist.  Otherwise `@user.votes` is `nil`.

```suggestion
    @user = User.find_by(id: params[:id])
    if @user.nil?
      redirect_to root_path
      flash[:error] = "User not found"
      return
    end
```

File: test/controllers/users_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      # Assert
+      must_respond_with :success
+    end
+
+    it "will show flash and render status for invalid id" do
+      # Act
+      get user_path(-1)
+
+      # Assert
+      must_respond_with :render
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb]: Render is not a response code.

```suggestion
      must_respond_with :redirect
```

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<h2>User Summary: <%=link_to "#{@user.name}", user_path(@user)%></h2>
+
+<p>Joined site <%="#{User.format_date(@user.created_at)}"%></p>
```
Comment (CheezItMan) [inline app/views/users/show.html.erb:3]: `format_date` isn't a method for `User`.  did you mean to make a custom method for the model?

You could add this method to `app/helpers/application_helper.rb` as a view helper:

```ruby
module ApplicationHelper
  def date_format(date)
    date.strftime("%d/%m/%Y %I:%M %p")
  end 
end
```

And then call it in your view:

```suggestion
<p>Joined site <%="#{date_format(@user.created_at)}"%></p>
```

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    it 'must have a title' do
+      work.title = nil
+
+      expect(work.valid?).must_equal false
+      expect(work.errors.messages).must_include :vin
```
Comment (CheezItMan) [inline test/models/work_test.rb]: I think you wanted to verify that the title wasn't valid. 

```suggestion
      expect(work.errors.messages.keys).must_include :title
```

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    <tr>
+      <th>Media Title</th>
+      <th>Created By</th>
+      <th>Published</th>
+      <th>Category</th>
+      <th>Voted On</th>
+    </tr>
+    </thead>
+    <tbody>
+    <% @user_votes.each do |vote| %>
```
Comment (CheezItMan) [inline app/views/users/show.html.erb:17]: You need to call `.votes` to get the collection of votes

```suggestion
    <% @user.votes.each do |vote| %>
```

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      <% work = Work.find_by(id: vote.work_id) %>
```
Comment (CheezItMan) [inline app/views/users/show.html.erb:18]: You can instead of using `find_by` here do `vote.work`

File: test/models/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+require "test_helper"
+
+describe User do
```
Comment (CheezItMan) [inline test/models/user_test.rb:3]: Just noting no real tests here.

File: test/models/vote_test.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+require "test_helper"
+
+describe Vote do
+  it 'can be instantiated' do
+    (Vote.new).must_be_kind_of Vote
```
Comment (CheezItMan) [inline test/models/vote_test.rb:5]: better to use a fixture

```suggestion
    expect(votes(:new_vote)).must_be_kind_of Vote
```

File: test/models/work_test.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
+require "test_helper"
+
+describe Work do
+
+  it 'can be instantiated' do
+    (Work.new).must_be_kind_of Work
```
Comment (CheezItMan) [inline test/models/work_test.rb]: This should test creating a work and that it's valid.

```suggestion
    work = works(:test_book)
    expect(work).must_be_kind_of Work
    expect(work.valid?).must_equal true
```

File: test/models/work_test.rb [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  it "will have the required fields" do
+    work = works(:test_book)
+    [:category, :title, :creator, :publication_year, :description].each do |check|
+      expect(work).must_respond_to check
+    end
+  end
+
+  describe "relationships" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:16]: No relationship tests.

File: test/models/work_test.rb [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    end
+
+    it 'must have a publication year' do
+
+    end
+
+    it 'must have a description' do
+
+    end
```
Comment (CheezItMan) [inline test/models/work_test.rb]: Just noting missing tests

File: test/models/work_test.rb [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  # Tests for methods you create should go here
+  describe "custom methods" do
+    # Your tests here
+  end
```
Comment (CheezItMan) [inline test/models/work_test.rb:70]: Just noting no custom method tests

File: test/controllers/users_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+  describe "show" do
+    it "can get a valid user" do
+          # Act
+          get user_path(first.id)
+
+      # Assert
+      must_respond_with :success
+    end
+
+    it "will show flash and render status for invalid id" do
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:34]: No check for the flash notice

File: test/controllers/users_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+      get user_path(-1)
+
+      # Assert
+      must_respond_with :render
+
+    end
+  end
+
+
+  describe "create" do
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:57]: Just noting no following tests.

File: test/controllers/votes_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,29 @@
+require "test_helper"
+
+describe VotesController do
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:3]: Just noting no tests

File: test/controllers/works_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+  describe 'new' do
+    it 'can get the works path' do
+      get new_work_path
+      must_respond_with :success
+    end
+  end
+  describe 'show' do
+    it 'can get the works path' do
+      get works_path
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb]: For the show action you need an id

```suggestion
      get work_path(Work.first.id)
```

File: test/controllers/works_controller_test.rb [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+      must_respond_with :success
+    end
+  end
+  # describe 'create' do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb]: Noting no following tests.

File: app/models/vote.rb [hunk 22]
```diff
@@ -0,0 +1,14 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :user_id, presence: true
+  validates :work_id, presence: true
+  validates :user_id, uniqueness: { scope: :work,
+                                    message: "Only one vote per work, please" }
```
Comment (CheezItMan) [inline app/models/vote.rb:8]: Awesome!

File: app/models/vote.rb [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+
+  def user_votes_table_row
```
Comment (CheezItMan) [inline app/models/vote.rb:10]: Clever

File: app/models/work.rb [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+  validates :title, presence: true, uniqueness: {scope: :category, case_sensitive: false},
+            length: {maximum: 40}
+  validates :category, presence: true, inclusion: {in: VALID_CATEGORIES,
+                                                   message: "%{value} is not a valid category"}
+  validates :creator, presence: true, length: {maximum: 30}
+  validates :publication_year, presence: true,
+            numericality: {only_integer: true,
+                           less_than_or_equal_to: Date.today.year,
+                           greater_than_or_equal_to: 1000}
+  validates :description, length: {maximum: 140}
```
Comment (CheezItMan) [inline app/models/work.rb:14]: 🏆

File: app/models/work.rb [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+      all_works[category] = Work.sort(works_by_category)
+    end
+    return all_works
+  end
+
+  def self.by_category(category)
+    Work.where(category: category).order(title: :asc).to_a
+  end
+
+  def vote_button(current_user)
```
Comment (CheezItMan) [inline app/models/work.rb:46]: Clever, you don't seem to be using it, but clever. 

I would suggest this would make a great [view helper](https://www.rubyguides.com/2020/01/rails-helpers/) since it has to do with how things look.

File: app/models/work.rb [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+
+  def print_number_of_votes
+    if self.votes.size != 1
+      return "#{self.votes.size} votes"
+    else
+      return "#{self.votes.size} vote"
+    end
+  end
+
+  def self.top_ten(sorted_hash)
```
Comment (CheezItMan) [inline app/models/work.rb:82]: This method is expecting the user of the method to call list all works.  I think you should just make this method require a category.  Then the method can call list_all works or sort internally and filter it by the category.  That way you can get the top ten in a category. 

I think you've just over-complicated the custom methods in this class.

Remember you can get all the "books" with:  `Work.where(category: "book")`  Then you can sort that list by the number of votes.

File: app/controllers/works_controller.rb [hunk 27]
```diff
@@ -0,0 +1,88 @@
+class WorksController < ApplicationController
+  before_action :find_work, only: [:show, :edit, :update, :destroy]
+  before_action :find_work_votes, only: :show
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:3]: Awesome!  Filters!

File: app/controllers/works_controller.rb [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+    if @work && @work.update(work_params)
+      flash[:success] = "Succesfully edited #{@work.category} \"#{@work.title}!\""
+      redirect_to work_path(@work.id)
+    elsif @work
+      flash.now[:error] = "Work edits not saved!"
+      @work.errors.messages.each do |field, messages|
+        flash[field] = messages
+      end
+      render :edit
+    end
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb]: If you're rendering use `flash.now`  Also don't forget the **status code**!

```suggestion
      flash.now[:error] = "Work edits not saved!"
      @work.errors.messages.each do |field, messages|
        flash.now[field] = messages
      end
      render :edit, status: :bad_request
    end
```

File: app/controllers/works_controller.rb [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+    if @work.save
+      flash[:success] = "Successfully created new #{@work.category} \"#{@work.title}!\""
+      redirect_to works_path
+    else
+      flash[:error] = "Work not created!"
+      @work.errors.messages.each do |field, messages|
+        flash[field] = messages
+      end
+      redirect_to new_work_path
+    end
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb]: You should instead render the new view.

```suggestion
      flash.now[:error] = "Work not created!"
      @work.errors.messages.each do |field, messages|
        flash.now[field] = messages
      end
      render :new, status: :bad_request
    end
```

File: app/controllers/works_controller.rb [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+  def destroy
+    if @work.nil?
+      @deleted_work = @work.destroy
+      flash[:success] = "#{@work.title} deleted"
+      redirect_to root_path
+    else
+      flash[:failure] = "Work #{@work.title} not destroyed."
+      redirect_back fallback_location: root_path
+    end
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb]: I think you want to destroy a work if it's **not** nil.

```suggestion
    unless @work.nil?
      @deleted_work = @work.destroy
      flash[:success] = "#{@work.title} deleted"
      redirect_to root_path
    else
      flash[:failure] = "Work #{params[:id]} not destroyed."
      redirect_back fallback_location: root_path
    end
```

File: config/routes.rb [hunk 31]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'works#home'
+
+  resources :works do
+    resources :votes, only: [:create, :destroy]
```
Comment (CheezItMan) [inline config/routes.rb:14]: I don't know that you need to nest `destroy`

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️, mostly, you don't have a movie category.  I'm uncertain as to why.
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️, I just can't vote if I'm not logged in.
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️, I can see the 2 categories you have
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️, you have front-end validations, nice, but I can't test the flash notice.  See my works controller notes.
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully | ⚠️ I can't delete a work!
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️, you don't present the vote button, but you have the validation as well. 
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ⚠️ Not quite there although you have a good start on it
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️, one minor note
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ⚠️, you have partials for forms but no others.  You can use partials to dry up the repeated lists of works you have in several views. 
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️, You have it, but you waaay overcomplicated the methods required. 
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ⚠️ Some improvement in the error handling and use of flash, and the destroy action.

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ⚠️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ⚠️, no real tests
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ⚠️ Very little testing
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ⚠, Some errors here especially around validations
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ⚠️ Missing tests

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | ♥️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅

## Summary

Yes, I'm labeling this Red, but you are closer than you think to Green.  Most of your issues relate to testing.  I've left extensive comments in the code.  Take a look.  You have a lot of the functionality in place. 

Areas to improve on include:

1.  Writing Validation and custom method tests for models
1.  Simplifying your custom model methods.  You overcomplicated your life with the `work.rb` file.  See my inline notes here.
1.  Practice debugging in Rails and asking for help.  I suspect your sharp brain let you do more with less help than other students in regular Ruby, but Rails is a bit bigger and you seemed to have run into a bug and moved on without calling out for help.  I'm happy to sit with you and review this project at some point and practice some debugging skills.

As time allows doing practice with controller tests as well.

Areas you did well:

1. Controller filters!  Awesome!  Well done.
1. Complicated validations, really clever ones
1. Front-end Validations
1. Lots of clever front-end things like hiding the vote button when I can't vote.
1. Pretty clean routes.rb file.
  
Due to the rubric this is a red, but you're actually pretty close to Green.  If testing had been done, I think it would have been Green.

File: config/initializers/constants.rb [hunk 32]
```diff
@@ -1 +1 @@
-VALID_CATEGORIES = %w[album book].freeze
+VALID_CATEGORIES = %w[album book movie].freeze
```
Comment (CheezItMan) [inline config/initializers/constants.rb:1]: 👍

File: test/controllers/users_controller_test.rb [hunk 33]
```diff
@@ -39,8 +39,20 @@
       must_respond_with :render
 
     end
-  end
 
+    describe "current" do
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:43]: What about users who are not logged in?

File: test/test_helper.rb [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+  # Add more helper methods to be used by all tests here...
+
+  def perform_login(user = nil)
+    user ||= User.first
+
+    login_data = {
+        user: {
+            username: user.name,
+        },
+    }
```
Comment (CheezItMan) [inline test/test_helper.rb:32]: Notice that the form you use to log in doesn't nest `username` under a `user` hash.  Also in your controller you are using "name" as the key and not "username".

So this should be:

```suggestion
    login_data = {
          name: user.name,
    }
```

File: config/routes.rb [hunk 35]
```diff
@@ -0,0 +1,20 @@
...
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'works#home'
+
+  get '/login', to: 'users#login_form', as: 'login'
+  post '/login', to: 'users#login'
+  delete '/logout', to: 'users#logout', as: 'logout'
+
+  get '/signup', to: 'users#new'
+
+  get "/users/current", to: "users#current", as: "current_user"
```
Comment (CheezItMan) [inline config/routes.rb:11]: You don't have this controller method.

File: test/controllers/users_controller_test.rb [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+
+    end
+
+    describe "current" do
+      it "returns 200 OK for a logged-in user" do
+        # Arrange
+        perform_login(users(:second))
+
+        # Act
+        get current_user_path
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:49]: This crashes because you don't have a current method in the users controller.

File: test/controllers/votes_controller_test.rb [hunk 37]
```diff
@@ -0,0 +1,94 @@
...
+        post work_votes_path(work)
+      }.wont_differ 'work.votes'
+
+      expect(flash[:error]).wont_be_nil
+    end
+    it 'cannot vote unless logged in' do
+      work = works(:test_book)
+      expect {
+        post work_votes_path(work)
+      }.wont_differ 'work.votes'
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:38]:
```suggestion
      }.wont_differ 'work.votes.count'
```

File: test/test_helper.rb [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+    post login_path, params: login_data
```
Comment (CheezItMan) [inline test/test_helper.rb:33]: Also a new user doesn't have an id yet so you may need to re-find that user in the database.

```suggestion
    post login_path, params: login_data
    user = User.find_by(name: user.name)
```

File: test/controllers/votes_controller_test.rb [hunk 39]
```diff
@@ -0,0 +1,94 @@
...
+  describe 'create' do
+    it 'can add a new vote' do
+      work = works(:test_album)
+      user = User.new(name: 'tester')
+
+      perform_login(user)
+
+      expect {
+        post work_votes_path(work)
+      }.must_differ 'work.votes', 1
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:13]: `must_differ` requires a number, so you need the count of the votes. 

```suggestion
      }.must_differ 'work.votes.count', 1
```

File: test/controllers/votes_controller_test.rb [hunk 40]
```diff
@@ -0,0 +1,94 @@
...
+
+      vote = Vote.find_by(user_id: user.id)
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:15]: Since you're creating a new user you'll also need to find them again in the DB since now they have an id number

```suggestion
       user = User.find_by(name: user.name)
      vote = Vote.find_by(user_id: user.id)
```

File: test/controllers/votes_controller_test.rb [hunk 41]
```diff
@@ -0,0 +1,94 @@
...
+              description: "John Leguizamo is the Pest"
+          }
+      }
+      post works_path, params: work_hash
+      user = User.new(name: 'tester')
+      perform_login(user)
+
+      expect {
+        delete work_vote_path(work_id: work_hash, vote_id: 1)
+      }.wont_differ 'work.votes'
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:79]: There is no work object here or vote for that matter and so it can't have votes.  This test doesn't make much sense.

File: test/controllers/votes_controller_test.rb [hunk 42]
```diff
@@ -0,0 +1,94 @@
...
+      work = works(:test_book)
+      expect {
+        post work_votes_path(work)
+      }.wont_differ 'work.votes'
+
+      expect(flash[:warning]).wont_be_nil
+    end
+  end
+
+  describe 'destroy' do
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:44]: For the destroy route, I don't think you need to use a nested route since you can identify the vote by it's id number.

File: test/controllers/votes_controller_test.rb [hunk 43]
```diff
@@ -0,0 +1,94 @@
...
+    it 'can remove an upvote' do
+      work = works(:test_album)
+      user = users(:first)
+
+      perform_login(user)
+      vote = Vote.find_by(user_id: user.id)
+
+      expect {
+        delete work_vote_path(work_id: work, vote_id: vote)
+      }.must_differ 'work.votes', -1
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:54]:
```suggestion
      expect {
        delete work_vote_path(work_id: work, id: vote)
      }.must_differ 'work.votes.count', -1
```

File: app/controllers/votes_controller.rb [hunk 44]
```diff
@@ -0,0 +1,47 @@
...
+      end
+    when "Remove Vote"
+      flash[:warning] = "Vote not created. You have already upvoted this work."
+    when nil
+      flash[:warning] = "Vote not created. You must be logged in to do that!"
+    end
+    redirect_to work_path(params[:work_id])
+  end
+
+  def destroy
```
Comment (CheezItMan) [inline app/controllers/votes_controller.rb:20]: I don't think you need a destroy or "downvote" method.  

I think you've overcomplicated this quite a bit.  If you want to give them the abiltiy to downvote, you can check to see if the current user and work have a vote, if so remove it, otherwise add a vote.

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully | ✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️, you don't present the vote button, but you have the validation as well. 
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️, one minor note
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ⚠️, you have partials for forms but no others.  You can use partials to dry up the repeated lists of works you have in several views. 
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️, You have it, but you waaay overcomplicated the methods required. 
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ⚠️ Some improvement in the error handling and use of flash, and the destroy action.

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ⚠️ Missing tests

## Overall Feedback

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 💚






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅

## Summary

This is green and it mostly works.  You do have a broken user show path, but that's minor.  I've also noted that your controller tests are a bit of a mess and tried to give some feedback there even though it wasn't in the assignment.  It's good that you attempted them, but I would get **one** test working before moving to the next.  Otherwise things tend to build like your last commit notes.  

As I noted last time, I think you overthought some things about the project, particularly the upvoting and downvoting logic you were working on.  Keep it simple.
